From: Hannes Reinecke <hare@suse.de>
Date: Wed, 17 Jul 2013 14:58:29 +0200
Subject: scsi: Do not retry invalid function error
References: bnc#809122
Patch-Mainline: n/a

Hitachi USP-V returns 'Invalid function' when the internal
staging mechanism encountered an error. These errors should
not be retried on another path.

Signed-off-by: Hannes Reinecke <hare@suse.de>

diff --git a/drivers/scsi/scsi_error.c b/drivers/scsi/scsi_error.c
index d101250..1f5aaed 100644
--- a/drivers/scsi/scsi_error.c
+++ b/drivers/scsi/scsi_error.c
@@ -467,6 +467,16 @@ static int scsi_check_sense(struct scsi_cmnd *scmd)
 			return TARGET_ERROR;
 
 	case ILLEGAL_REQUEST:
+		if (!strncmp(scmd->device->vendor, "HITACHI", 7) &&
+		    !strncmp(scmd->device->model, "OPEN-V", 6) &&
+		    (sshdr.asc == 0x22)) {
+			/*
+			 * Hitachi USP-V returns 'Invalid function'
+			 * when an internal error occurred; this
+			 * error should not be retried on other paths
+			 */
+			return TARGET_ERROR;
+		}
 	default:
 		return SUCCESS;
 	}
